### DISCUSSION TITLE
v3::sink::send_at_least_once() and message ordering

### DISCUSSION THREAD

**Discussion by Antiarchitect** [ntex-rs/ntex-mqtt#82] [Q&A]
Suppose I'm receiving messages to be sent to the client from the channel in a loop and there:

```
loop {
   ...
   rx.recv_async().await {
       if sub.qos == QoS::AtMostOnce {
          builder.send_at_most_once();
       } else {
          ntex::rt::spawn(builder.send_at_least_once());
       };
   }
   ...
}

```

Is there any guarantee that messages are sent to the client in the same order as they received from rx or I should maintain the order by waiting `ntex::rt::spawn(builder.send_at_least_once())` each time?

**Comment by fafhrd91** [comment]
there are no strong guarantees but messages should be send in enqueue order

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Andrey Voronkov ***@***.***>
Sent: Tuesday, December 21, 2021 6:28:04 AM
To: ntex-rs/ntex-mqtt ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [ntex-rs/ntex-mqtt] v3::sink::send_at_least_once message ordering (Discussion #82)


Suppose I'm receiving messages to be sent to the client from the channel in a loop and there:

loop {
   ...
   rx.recv_async().await {
       if sub.qos == QoS::AtMostOnce {
          builder.send_at_most_once();
       } else {
          ntex::rt::spawn(builder.send_at_least_once());
       };
   }
   ...
}



Is there any guarantee that messages are sent to the client in the same order as they received from rx or I should maintain the order by waiting ntex::rt::spawn(builder.send_at_least_once()) each time?

—
Reply to this email directly, view it on GitHub<https://github.com/ntex-rs/ntex-mqtt/discussions/82>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAESHTWXCRVOOEMWCB2VVY3UR7CZJANCNFSM5KO5MT4Q>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
